File: macos/Sources/VoiceKeyCore/CueSound.swift
import Foundation

/// One note of a cue: a pitch, held for a while.
public struct CueTone: Equatable {
    public let hertz: Double
    public let milliseconds: Int

    public init(hertz: Double, milliseconds: Int) {
        self.hertz = hertz
        self.milliseconds = milliseconds
    }
}

/// The two sounds that say VoiceKey started and stopped listening. Synthesised
/// rather than shipped as files, for the same reason the menu-bar glyph is drawn
/// in code: nothing to lose, nothing to load, and the pitches are readable here.
///
/// The pair is one interval played both ways round — E5 up to B5 to start, back
/// down to stop — so which one you heard needs no learning.
public enum CueSound {

    public static let sampleRate = 48_000

    /// Well under full scale: this sits behind whatever the user is doing.
    private static let amplitude = 0.35

    private static let fadeInSamples = sampleRate * 4 / 1000
    private static let fadeOutSamples = sampleRate * 20 / 1000

    private static let lowNote = 659.25   // E5
    private static let highNote = 987.77  // B5

    public static let start = [CueTone(hertz: lowNote, milliseconds: 70),
                               CueTone(hertz: highNote, milliseconds: 90)]

    public static let stop = [CueTone(hertz: highNote, milliseconds: 70),
                              CueTone(hertz: lowNote, milliseconds: 90)]

    /// The sidebar footer row that turns the cues off, worded like the
    /// launch-at-login row beside it: the state is the word after the dash.
    public static func label(_ enabled: Bool) -> String {
        "Sounds — \(enabled ? "On" : "Off")"
    }

    /// The tones as a canonical 16-bit mono PCM WAV, header and all.
    public static func wav(_ tones: [CueTone]) -> [UInt8] {
        let samples = tones.flatMap(render)
        let dataBytes = samples.count * 2

        var wav: [UInt8] = []
        wav.reserveCapacity(44 + dataBytes)
        wav += Array("RIFF".utf8)
        wav += bytes(Int32(36 + dataBytes))
        wav += Array("WAVE".utf8)
        wav += Array("fmt ".utf8)
        wav += bytes(Int32(16))                       // PCM header length
        wav += bytes(Int16(1))                        // PCM, uncompressed
        wav += bytes(Int16(1))                        // mono
        wav += bytes(Int32(sampleRate))
        wav += bytes(Int32(sampleRate * 2))           // bytes per second
        wav += bytes(Int16(2))                        // bytes per frame
        wav += bytes(Int16(16))                       // bits per sample
        wav += Array("data".utf8)
        wav += bytes(Int32(dataBytes))
        wav += samples.flatMap { bytes($0) }
        return wav
    }

    private static func render(_ tone: CueTone) -> [Int16] {
        let count = sampleRate * tone.milliseconds / 1000
        return (0..<count).map { index in
            Int16(sin(2 * Double.pi * tone.hertz * Double(index) / Double(sampleRate))
                  * amplitude * fade(index, count) * Double(Int16.max))
        }
    }

    /// Silent at both ends of every tone. A note that starts or stops mid-cycle
    /// is a step in the waveform, and a step is the click we are avoiding.
    private static func fade(_ index: Int, _ count: Int) -> Double {
        let rise = min(1.0, Double(index) / Double(min(fadeInSamples, count)))
        let fall = min(1.0, Double(count - 1 - index) / Double(min(fadeOutSamples, count)))
        return min(rise, fall)
    }

    private static func bytes<T: FixedWidthInteger>(_ value: T) -> [UInt8] {
        withUnsafeBytes(of: value.littleEndian, Array.init)
    }
}

File: windows/VoiceKey.Core/CueSound.cs
namespace VoiceKey.Core;

/// <summary>One note of a cue: a pitch, held for a while.</summary>
public readonly record struct CueTone(double Hertz, int Milliseconds);

/// <summary>
/// The two sounds that say VoiceKey started and stopped listening. Synthesised
/// rather than shipped as files, for the same reason the tray glyph is drawn in
/// code: nothing to lose, nothing to load, and the pitches are readable here.
///
/// The pair is one interval played both ways round — E5 up to B5 to start, back
/// down to start again — so which one you heard needs no learning.
/// </summary>
public static class CueSound
{
    public const int SampleRate = 48_000;

    /// <summary>Well under full scale: this sits behind whatever the user is doing.</summary>
    private const double Amplitude = 0.35;

    private const int FadeInSamples = SampleRate * 4 / 1000;
    private const int FadeOutSamples = SampleRate * 20 / 1000;

    private const double LowNote = 659.25;  // E5
    private const double HighNote = 987.77; // B5

    public static IReadOnlyList<CueTone> Start { get; } =
        [new CueTone(LowNote, 70), new CueTone(HighNote, 90)];

    public static IReadOnlyList<CueTone> Stop { get; } =
        [new CueTone(HighNote, 70), new CueTone(LowNote, 90)];

    /// <summary>
    /// The sidebar footer row that turns the cues off, worded like the launch-at-login
    /// row beside it: the state is the word after the dash.
    /// </summary>
    public static string Label(bool enabled) => $"Sounds — {(enabled ? "On" : "Off")}";

    /// <summary>The tones as a canonical 16-bit mono PCM WAV, header and all.</summary>
    public static byte[] Wav(IReadOnlyList<CueTone> tones)
    {
        var samples = tones.SelectMany(Render).ToArray();
        var dataBytes = samples.Length * sizeof(short);

        using var stream = new MemoryStream(44 + dataBytes);
        using var writer = new BinaryWriter(stream);
        writer.Write("RIFF"u8);
        writer.Write(36 + dataBytes);
        writer.Write("WAVE"u8);
        writer.Write("fmt "u8);
        writer.Write(16);                          // PCM header length
        writer.Write((short)1);                    // PCM, uncompressed
        writer.Write((short)1);                    // mono
        writer.Write(SampleRate);
        writer.Write(SampleRate * sizeof(short));  // bytes per second
        writer.Write((short)sizeof(short));        // bytes per frame
        writer.Write((short)16);                   // bits per sample
        writer.Write("data"u8);
        writer.Write(dataBytes);
        foreach (var sample in samples) writer.Write(sample);

        writer.Flush();
        return stream.ToArray();
    }

    private static IEnumerable<short> Render(CueTone tone)
    {
        var count = SampleRate * tone.Milliseconds / 1000;
        return Enumerable.Range(0, count).Select(index =>
            (short)(Math.Sin(2 * Math.PI * tone.Hertz * index / SampleRate)
                    * Amplitude * Fade(index, count) * short.MaxValue));
    }

    /// <summary>
    /// Silent at both ends of every tone. A note that starts or stops mid-cycle
    /// is a step in the waveform, and a step is the click we are avoiding.
    /// </summary>
    private static double Fade(int index, int count)
    {
        var rise = Math.Min(1.0, index / (double)Math.Min(FadeInSamples, count));
        var fall = Math.Min(1.0, (count - 1 - index) / (double)Math.Min(FadeOutSamples, count));
        return Math.Min(rise, fall);
    }
}
